### ISSUE TITLE
feat(atlas-model): BF16 inference path — OLMo-3-7B-Think GPU throughput (4.1 → ~20 tok/s)

### ISSUE THREAD

**Issue by web3guru888** [web3guru888/ATLAS#9]
## Problem

OLMo-3-7B-Think currently runs at **4.1 tok/s on CPU** despite the server having an NVIDIA A100-SXM4-40GB (40GB VRAM). The bottleneck is a software architecture issue, not hardware.

### Root cause chain

1. `load_model_from_dir()` reads safetensors (BF16, ~14.6 GB on disk) and converts every tensor to **f32** on load → ~28 GB in system RAM
2. `GpuMatrix` attempts to upload f32 weights to VRAM — but 28 GB f32 either exceeds the safe VRAM allocation (40 GB minus runtime/activation headroom) or silently falls back
3. `forward_one_gpu()` returns `None` → CPU fallback for every token
4. Evidence: GPU VRAM stays at ~0 MiB during OLMo-3 inference; SmolLM2-135M (507 MiB) correctly uses GPU at 37.7 tok/s

---

### Option A — BF16 (recommended, implement first)

Keep weights at BF16 (2 bytes/param) rather than converting to f32 at load time.

- OLMo-3-7B in BF16: **~14 GB VRAM** — comfortably fits in 40 GB A100 with room for KV cache and activations
- Quality vs f32: **< 0.01% perplexity difference** — BF16 is the native training format, essentially lossless
- Expected throughput: **~15–25 tok/s** on A100 (BF16 tensor cores)
- This is what every production inference stack does (vLLM, llama.cpp, HuggingFace generate)

**Implementation sketch:**

```rust
// atlas-tensor: add BF16 GpuMatrix variant
pub enum GpuMatrix {
    F32 { ptr: GpuBuf<f32>, rows: usize, cols: usize },
    BF16 { ptr: GpuBuf<u16>, rows: usize, cols: usize },  // NEW
}

// atlas-model: keep BF16 tensors from safetensors instead of converting
// Current:  bf16_slice → f32 vec → GpuMatrix::F32
// Proposed: bf16_slice → GpuMatrix::BF16 (upload directly)

// kernels/matmul.cu: add BF16 SGEMM path
// A100 natively supports __nv_bfloat16 (cuda_bf16.h, CUDA 11+)
__global__ void sgemm_bf16(...) { ... }
```

Changes needed:
- `atlas-tensor`: `GpuBuf<u16>` + BF16 upload/download + `sgemm_bf16` CUDA kernel
- `atlas-model`: skip f32 conversion when loading safetensors if BF16 path available; `Linear::forward_gpu` dispatches to `sgemm_bf16`
- `atlas-model`: RMSNorm, RoPE, SwiGLU kernels need BF16 variants (or promote to f32 for activations — W16A32 is fine)

### Option B — INT8 W8A16 (implement second, after BF16 baseline)

Quantize weights to INT8 at load time, keep activations at f16.

- OLMo-3-7B in INT8: **~7 GB VRAM** — fits easily, leaves 33 GB headroom
- Quality vs BF16: **~1–3% accuracy drop** on reasoning benchmarks (MMLU, GSM8K)
- OLMo-Think is a chain-of-thought reasoning model — quantization errors can corrupt thinking traces on hard problems. W8A16 is the right balance; W8A8 risks more visible degradation.
- Expected throughput: **~35–40 tok/s** on A100 (~2× over BF16)
- Implement *after* BF16 baseline so quality delta can be measured empirically

| Scheme | VRAM | vs f32 quality | A100 tok/s (est.) |
|--------|------|---------------|-------------------|
| f32 (current) | ~28 GB | baseline | 4.1 (CPU fallback) |
| **BF16 (Phase 1)** | **~14 GB** | **< 0.01% loss** | **~20 tok/s** |
| INT8 W8A16 (Phase 2) | ~7 GB | ~1–3% loss | ~35–40 tok/s |
| INT4 GPTQ/AWQ | ~4 GB | ~5–8% loss | ~50–60 tok/s |

---

## Suggested implementation order

1. **BF16 GpuBuf + upload kernel** in atlas-tensor — store `u16` on GPU, convert to f32 in kernel (W16A32)
2. **Skip f32 conversion** in `load_model_from_dir` when source tensor is BF16
3. **BF16 SGEMM kernel** in `kernels/matmul.cu` using `__nv_bfloat16` (CUDA 11+, A100 natively supported)
4. Benchmark BF16 vs current f32-CPU baseline on OLMo-3-7B-Think
5. If BF16 throughput is sufficient, stop. If higher throughput needed, add INT8 W8A16 path using existing `atlas-quant` INT8 infrastructure

---

## Hardware context

- Server: A100-SXM4-40GB, CUDA 12.9, sm_80, 85 GB RAM, 12 vCPU
- No hardware upgrade needed — A100 has native BF16 tensor cores (312 TFLOPS BF16 vs 77 TFLOPS f32)
- The 4.1 tok/s figure is the CPU RAM-bandwidth ceiling for f32; GPU would be ~5–7× faster

## Related

- Closes part of the performance gap surfaced by #7 (OLMo-3-7B now produces correct output, but slowly)
- Existing `atlas-quant` crate has INT8 symmetric scaling infrastructure that can be reused for Phase 2
- `GpuMatrix` pre-upload pattern (already used for SmolLM2) is the right abstraction — just needs BF16 variant

**Comment by web3guru888** [comment]
## ✅ Resolved — BF16 GPU Inference Path Live

**Final result: 19.9 tok/s on A100-SXM4-40GB** — 4.8× speedup over the 4.1 tok/s CPU baseline.

---

### What was implemented (3 commits)

#### `e7986cc` — feat(atlas-model, atlas-tensor): BF16 GPU inference path
- **`SafetensorsFile::get_bf16()`** — reads BF16 tensors as raw `Vec<u16>` bit patterns (no precision loss)
- **`GpuBufBf16`** struct — cudaMalloc/cudaMemcpy/cudaFree for `*mut u16` buffers (2 bytes/elem vs 4)
- **`GpuBufKind`** enum — `{ F32(GpuBuf), BF16(GpuBufBf16) }` — GpuMatrix stores either precision
- **`GpuMatrix::upload_bf16()`** — uploads BF16 weights directly to VRAM halving memory footprint: 28 GB f32 → 14 GB bf16
- **`atlas_sgemm_bf16_f32()`** CUDA wrapper — BF16 weights × f32 activations → f32 output
- **BF16 tiled GEMM kernel** for batch/prefill (N > 1)
- `load_model_from_dir()` now tracks BF16 tensors via `HashSet` and routes to `make_linear_bf16_aware()`
- `OlmoModel::gpu_weight_dtype_counts()` — diagnostic: reports (bf16_count, f32_count) across all layers

#### `af9eda1` — perf(kernels): add GEMV kernels for N=1 — fix 32× inefficiency in autoregressive decode
- **Root cause of 4.4 tok/s (barely above CPU)**: the existing tiled GEMM with TILE=32 for N=1 only activates threadIdx.x=0 to write output — 31/32 threads multiply against `sB[k][threadIdx.x]=0.0f` → 97% compute wasted, kernel becomes compute-bound not memory-bandwidth-bound
- **Fix**: dedicated `sgemv_f32_kernel` and `sgemv_bf16_kernel` — one warp (32 threads) per output row, warp-reduction via `__shfl_down_sync`, `__ldg` on the activation vector
- Dispatch logic: both `atlas_matmul_f32` and `atlas_sgemm_bf16_f32` check `if (N == 1)` and route to GEMV kernels
- Result: memory-bandwidth-limited as expected for GEMV on HBM2e

#### `f2f723b` — fix(kernels): move bf16u_to_f32 before GEMV kernels that use it
- nvcc forward-declaration fix: `bf16u_to_f32` helper moved to its own section before the kernels that call it

---

### Verified benchmark output (A100-SXM4-40GB, release build)

```
[load_model_from_dir] loaded 355 tensors total (355 BF16 → GPU BF16 path)
  GPU matrices: 224 BF16 + 0 F32 / 224 total
  GPU logits: min=0.266 max=45.053 spread=44.787
  OLMo-3-7B BF16 GPU  | load=109s  | 1st=33ms | 19.9 tok/s
  Speedup vs CPU baseline (4.1 tok/s): 4.8×
```

Test: `gpu_benchmark_olmo3_7b_think_bf16` ✅  
Parity test: `gpu_bf16_gemv_parity` ✅  
Full suite: **528 tests, 0 failures** ✅

---

### Architecture note — W16A32

The implementation follows the W16A32 pattern (weights in BF16, activations in f32). This halved VRAM from ~28 GB (f32) to ~14 GB (bf16), allowing all 224 weight matrices to reside in the A100's 40 GB HBM2e. The BF16→f32 conversion inside the GEMV kernel (`__uint_as_float(((uint32_t)h) << 16)`) has zero precision loss since BF16 is exactly the upper 16 bits of an IEEE 754 f32 bit pattern.
